Zend/zend_execute.c

```diff
@@ -1896,8 +1896,12 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+			} else if (EXPECTED((EG(current_execute_data)->opline->opcode == ZEND_FETCH_LIST && Z_LVAL_P(dim) == 0 && Z_TYPE_P(container) != IS_FALSE && (Z_TYPE_P(container) != IS_STRING || (Z_TYPE_P(container) == IS_STRING && Z_STRLEN_P(container) != 0))) || (EG(current_execute_data)->opline->opcode != ZEND_FETCH_LIST && (EG(current_execute_data)->opline->op1_type != IS_VAR || (Z_TYPE_P(container) != IS_UNDEF && Z_TYPE_P(container) != IS_NULL && (support_strings && Z_TYPE_P(container) != IS_STRING)))))) {
```

Comment:
I'm not sure special-casing each() is a good idea. At least, the way you do it (not specifically tailored to each()) will have side-effects on other cases as well (and tailoring it is even more crazy). This condition looks like a large assembly of various special-cases, which will introduce a bunch of extra inconsistencies resulting in different behavior based on minor differences in code -- like whether something reads a simple variable or an array offset/function call result.

(I'd actually consider it a benefit if uses of each() in that started to generate notices -- serves as a natural deprecation. This pattern needs to die.)